### PR TITLE
Add room querying functions

### PR DESCRIPTION
**Pull request by nihey**
As requested in https://github.com/miguelgrinberg/Flask-SocketIO/issues/105

- room function will return the list of currently connected clients to a room.
- rooms function will return the name of all rooms.

### CODE REVIEW

**Comment by nihey** [comment]
I made a observation on my code, if you feel it's ok, I could proceed and add some tests on the next patch.

File: flask_socketio/__init__.py [hunk 1]
```diff
@@ -70,6 +70,15 @@ def process_event(self, packet):
                 app = self.request
                 return self.socketio._dispatch_message(app, self, message, args)
 
+            def get_room(self, room):
+                if not self.socketio.rooms.get(self.ns_name, False):
+                    return []
+                return list(self.socketio.rooms[self.ns_name].get(room, []))
```
Comment (miguelgrinberg) [inline flask_socketio/__init__.py]: So this was one of my concerns. This function is returning a private object maintained by gevent-socketio, a project that has not been maintained in a long time.

I'm hoping to replace gevent-socketio in the future, so returning this private structure makes my life harder, as I will then be forced to continue using the gevent-socketio structures, or else break current users when I switch to the new thing.

My preference would be to abstract the application from these internal structures. Here you could return a list of client IDs for example, and then add new option to the emit and send methods to take this ID to direct a message to a specific user.

Comment (nihey) [inline flask_socketio/__init__.py]: I'm sorry, I thought:

> But I would be okay with exposing the list of rooms and the list of clients in a room through functions that I can later reimplement (...)

Meant to place some stubs with the old structures that could later be reimplemented.

But that's alright, I can workaround using the proposed `socket.sessid`. I'll be back with another patch.

**Comment by miguelgrinberg** [comment]
I'm still thinking about this, because I'm starting work on a rewrite of this extension in a way that is Python 3 compatible, without using gevent-socketio or gevent.

Unfortunately there are situations I envision where obtaining the list of clients in a room might be impractical or expensive. Consider the use of multiple workers, where each worker handles a subset of the connections. In this scenario returning the list of clients is pretty hard, as it may require transferring large lists between multiple worker processes.

So my question is: wouldn't it make more sense to track room assignments at the application level? The application can use a database for this, updated every time a connect or disconnect event is triggered. Also, as I mentioned earlier, from the Socket.IO perspective, a client is just represented by a session ID and a namespace, so there isn't really much that the application can use. If this is done by the application, then users can be identified and authenticated, and then the association between rooms and useres can be tracked directly instead of through these abstract sessions and namespaces that are specific to the Socket.IO protocol.

**Comment by nihey** [comment]
> wouldn't it make more sense to track room assignments at the application level?

Even though I don't have strong knowledge in asynchronous python application. I think this makes sense. At least it would be much more practical and readable.

**Comment by ktoh** [comment]
As an aside I agree that, at minimum, the gevent socketio dependency is really limiting growth of this (great) plug in because of how stale that library is. Look forward to hearing about the outcome of your thought process!

**Comment by miguelgrinberg** [comment]
This functionality will not be supported, as it is incompatible with supporting multiple worker processes, a feature that is soon going to be implemented.

**Pull request closed without merging**